core/Updates/3.0.0-b2.php

```diff
@@ -21,4 +23,84 @@ public static function isMajorUpdate()
     {
         return true;
     }
+
+    /**
+     * @var MigrationFactory
+     */
+    private $migration;
+
+    /**
+     * @var string
+     */
+    private $userTable = 'user';
+
+    public function __construct(MigrationFactory $factory)
+    {
+        $this->migration = $factory;
+    }
+
+    /**
+     * Here you can define one or multiple SQL statements that should be executed during the update.
+     * @param Updater $updater
+     * @return Migration[]
+     */
+    public function getMigrations(Updater $updater)
+    {
+        $migrations = [];
+        $migrations = $this->getUserDatabaseMigrations($migrations);
+
+        return $migrations;
+    }
+
+
+    public function doUpdate(Updater $updater)
+    {
+        $updater->executeMigrations(__FILE__, $this->getMigrations($updater));
+
+        $this->hashUserPasswords();
+    }
+
+      /**
+     * @param Migration[] $queries
+     * @return Migration[]
+     */
+    private function getUserDatabaseMigrations($queries)
+    {
+        $queries[] = $this->migration->db->changeColumn($this->userTable, 'password', 'password', 'VARCHAR(255) NOT NULL');
+        $queries[] = $this->migration->db->addColumn($this->userTable, 'password_hashed', 'VARCHAR(255) NOT NULL', 'password');
```

Comment:
I think `getTokenAuth` and `generateTokenAuth` require the separation you mentioned already for being able to include any random data. Before accessing the token the authentication should have already happened so it should be perfectly fine to just access the user record and return what is found there.

The generation API might still need some sort of public interface as with random generation there is no need to reference the password anymore. So the token could be generated completely independent from password changes. The first password change should automatically do the migration away from the old MD5 token with an updated warning in the interface.

Regarding the passwords I also thought about just applying bcrypt around already hashed MD5 passwords. Besides rendering the legacy option obsolete it would ensure full backwards compatibility without the current "hack".

It leaves some of the problems with MD5 in place like collision vulnerability or the fact that bcrypt now works on a fixed length hash instead of potential 72 chars. But that is already a problem due to the API requirements for the MD5 login. At least it made the problem obvious :)